NangoHQ/nango
# feat(shared): Dynamic reload for providers

## Describe your changes

Adds the ability to reload providers dynamically. This is handled by a new function, `launchProvidersSync` that uses `setTimeout` in the background to monitor for updates to the providers file.

To utilize a provider's file over http, set the `PROVIDERS_URL` environment variable. Optionally set `PROVIDERS_RELOAD_INTERVAL` to a number of milliseconds to wait between reload attempts.

- [x] Add launchProvidersSync to services that need it
- [x] Log when providers data updates so that it's clear
- [x] Add ability to load providers over HTTP
- [x] Make the reload frequency configurable

### How to review it

1. Set up a github gist with the content of `providers.yaml`.
2. Add `/raw` to the end of the gist URL to get the raw URL for it. Set that value to `PROVIDERS_URL` in `.env`.
3. Run the app: `npm run dwa`
4. Update the gist. You'll see a log message within the reload frequency time showing that it's been updated.
5. Run a runner script that takes longer than the reload frequency and update the gist while it's running to see a log message saying it has been updated.

### Open Questions

- What do we want the default reload frequency to be? I currently have it set to 30 seconds.
- Is it dangerous to just replace the providers while the app is running? Based on cusory code review it doesn't seem to be, but it'd be helpful to have more experienced feedback to that problem.

### Deployment Notes

- Do we want to simply use the `providers.yaml` from GitHub or do we want to host a specific version somewhere?
- We'll need to set up an ENV var for `PROVIDERS_URL`. If that env var isn't set, the app will work like it did before.

### To Clean Up

- Right now I'm putting a param onto the url since I'm hosting with a gist and need to be able to cache bust. We may want to keep this if we go with the `providers.yaml` out of the repo. I did notice with a gist that it flapped between versions at least once - something to watch for with however we host the file.

### Things I thought about and didn't do

- It doesn't feel like reloads from disk are important to this project since our main goal is to allow for reloads without restarts, and upgrading the default `providers.yaml` would probably happen along with a package update.
- It also doesn't feel like allowing someone to configure a disk path for `providers.yaml` is that important to the current project.
- I didn't write any tests because there were no tests for these files to start with and they'd be better as integration tests but are still pretty likely to be flaky. More than happy to dig into testing this more, though.

### Things I noticed

- I had to run `npm run ts-build` to get `shared` to rebuild. `npm run dwa` didn't seem to pick up code changes there.
- The code feels like we intend to just load `providers.yaml` from the module by default. Would it make sense to remove the file resolution code in `getProvidersPath` and just try to load `providers.yaml` from the default location?

## Issue ticket number and link

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:
